developer_guides/dsp_core/index.rst

```diff
@@ -0,0 +1,81 @@
+.. _dsp_core:
+
+DSP Cores Management
+####################
+
+More and more audio DSPs with multiple cores are implemented in the current
+SoC(System on Chip), to fully ultilize the bandwidth of all cores, and do good
+power saving with power on/off cores dynamically, we need a generic DSP cores
+management with SOF stack.
+
+Context -- pipeline, component, and DSP core
+********************************************
+
+Driver pipeline and widget
+==========================
+
+A pipeline usually contains one scheduler widget (or named 'pipeline widget in
+the FW') + several other widgets, all these widgets are 'swidget's in the
+driver side.
+
+Firmware pipeline, component and DSP core
+=========================================
+
+In the firmware side, each pipeline has a dedicated pipeline task(related to
+pipeline_copy()), to be scheduled on a specified DSP core (set in topology via
+scheduler widget's core).
+
+Each component (normal driver widget) can have its own task also, it can be
+specified to run on a specified DSP core, this is not neccesary to be the same
+one with the pipeline one.
+
+
+How to manage the DSP cores in the driver?
+******************************************
+
+- use a refcount for each DSP core to denote if there is any task ask for
+  using it.
+- increase the refcount via _get() each time a widget (its task actually)
+  ask for it, in widget_setup(), before the widget/component can be created
+  via IPC component_new.
+- decrease the refcount via _put() after a widget is freed, in widget_free().
+- power on a DSP core at the point the refcount is changing from 0 to 1, and
+  power off it at 1 to 0.
+
+How to power on/off a DSP core?
+*******************************
+
+To power on a DSP core, we need to
+
+- power on it in the driver side (if the core is host manged) usually via
+  register setting,
+- then send an core_enable IPC (with the core bit masked) to the FW to ask
+  for DSP core setting up in the FW side.
+
+To power off a DSP core, we need to
+
+- send an core_enable IPC (with the core bit unmasked) to the FW to ask for
+  DSP core tearing down in the FW side,
+- power off it in the driver side (if the core is host manged) usually via
+  register setting.
+
+The life cycle of a DSP core
+****************************
+
++-------------+---------+--------------+---------------------------+----------------------------+----------------------+
+| Cores/stage | booting	|    booted    |	widget setup	   |       widget free		|	suspend        |
++=============+=========+==============+===========================+============================+======================+
```

Comment:
This is very dynamic pipeline and topology specific. What we want is more of how a user (not just topology) should use the multi-core implementation. ie what APIs to use for requesting a core and releasing it when no longer needed. For example, the FW booting is simply a user rather than a stage in the DSP core's life. 

we are tying 2 separate concepts here and I think our main focus should be how cores are managed by the driver and firmware rather than how dynamic pipelines should use the cores.

Reply:
> This is very dynamic pipeline and topology specific. What we want is more of how a user (not just topology) should use the multi-core implementation. ie what APIs to use for requesting a core and releasing it when no longer needed. For example, the FW booting is simply a user rather than a stage in the DSP core's life.

Yep, that sounds more like a user guide for driver writing, sounds good.

> 
> we are tying 2 separate concepts here and I think our main focus should be how cores are managed by the driver and firmware rather than how dynamic pipelines should use the cores.

I added this table, with intention to provide a full picture about all the users who will touch the DSP cores power change, and a typical sequence about how they interact with each other.